File: 0611-valid-triangle-number/0611-valid-triangle-number.cpp
class Solution {
public:
    int triangleNumber(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int count = 0;

        // 固定最大邊 nums[k]
        for (int k = n - 1; k >= 2; k--){
            int i = 0, j = k - 1;
            while (i < j){
                if (nums[i] + nums[j] > nums[k]){
                    count += (j - i); // i 到 j-1 都成立
                    j--;
                } else{
                    i++;
                }
            }
        }
        return count;
    }
};

File: 0611-valid-triangle-number/0611-valid-triangle-number.py
class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        n = len(nums)
        count = 0

        for k in range(n - 1, -1, -1): # 固定最大邊
            i, j = 0, k - 1
            while i < j:
                if nums[i] + nums[j] > nums[k]:
                    count += (j - i) # 所有 i...j-1 都可以
                    j -= 1
                else:
                    i += 1
        return count